### ISSUE TITLE
[feature request] chdir() into an archive

### ISSUE THREAD

**Issue by cdluminate** [fish-shell/fish-shell#5166]
It would make fish more different from bash and zsh if fish user can `cd` into a compressed archive. I wanted this feature several years ago and I wrote [this tool](https://github.com/cdluminate/cda) but it's still hacky. Apart from really extracting this archive, avfs looks like a good solution too.

Thank you fish devs for the lovely fish shell.

**Comment by zanchey** [comment]
My feeling is that this is out of scope, although I'm keen to hear other opinions. There's no way of implementing this without extracting the full archive somewhere, which is difficult to do cleanly using out-of-process tools and dangerous to do in-process.

**Comment by mqudsi** [comment]
Not commenting on whether it _should_ be done, but in theory it is certainly possible to cd into a tarball without first extracting it to a temporary location by faking the navigation from the tarball index. OP's suggestion is to use a fuse plugin that would mount the tarball to a temporary folder abstracting away that directory emulation and allowing fish to treat it like a real folder.

**Comment by zanchey** [comment]
The problem there is external commands like `cp`.

**Comment by krader1961** [comment]
The only way to make this work is to unpack the archive (which is problematic) or use FUSE (or a similar mechanism) to mount it as a read-only filesystem so that normal system APIs (chdir, open, stat, etc.) work. Not just for fish builtins but also any external command that might be run.

**Comment by faho** [comment]
This is an interesting idea, but one I'm pretty firmly against.

To do this right, as said, this would have to be transparent to external commands or the experience is awful (e.g. if you used globs, they would have to actually _work_ with `ls`, and `cp` et al).

That would entail using something _like_ FUSE, which may or may not be available on all operating systems fish is used on. We'd then have to add that as a dependency or (even worse) maintain it ourselves.

And what use would this actually be? You couldn't `rm` a file from an archive, editting files would presumably be impossible (since I don't think we'd want transparent recompression), `grep` and `cat` would be possible but `zgrep` and `zcat` already exist. And obviously just unpacking the file is always an alternative to using this functionality, which IMHO makes the gains pretty small.

**Comment by mqudsi** [comment]
I think the *only* implementation of this that wouldn't suck would be if fish had a syntax to grab files from within an archive, e.g. `grep hello !foo.tar.gz:/some/file` where `!___:___` would be understood to mean "extract this file from this archive to a temporary location and replace this argument with its path", but that means new syntax which is a) undiscoverable, b) a compatibility nightmare and something we're actively running away from.

You can't just use a blind "foo.tar.gz:some/path" because that could possibly be something you want passed as-is to a command, unless you would want to require _that_ to always be quoted, etc.

Basically, it's a real pita any way you look at it.

**Comment by zanchey** [comment]
I don't think we're taking this - sorry.

**Comment by cdluminate** [comment]
Then I think avfs would be a good way to implement this feature, and fish need to do nothing about archive files.

https://sourceforge.net/p/avf/git/ci/master/tree/README#l86
```
'cd' into an archive:
  (For the 'preload' method the shell itself must be started with
   avfs)

  cd avfs-0.9.1.tgz#/
```

We just need add a wrapper around `cd` and update `cd`'s completion to accept archives as candidates when avfs is available. The wrapper just need to lead the user to the avfs mounted directory when the user tries to cd into a file.
